File: src/logic/voteRecords.js
// EPIC-07 投票済みを記録できる。
import { loadState, saveState, nextId, nowIso } from "./db.js"
import { getElection } from "./elections.js"
import { getRegion } from "./regions.js"

// 「投票した」と記録する。同じ選挙への記録は冪等（二重記録しない）。
export function markVoted(regionId, electionId) {
  const state = loadState()
  getRegion(regionId, state) // 存在確認
  getElection(electionId, state) // 存在確認
  let record = state.vote_records.find((r) => r.region_id === regionId && r.election_id === electionId)
  if (!record) {
    record = {
      id: nextId(state, "vote_records"),
      region_id: regionId,
      election_id: electionId,
      voted_at: nowIso(),
    }
    state.vote_records.push(record)
    saveState(state)
  }
  return record
}

// 投票済みの選挙・過去の投票履歴を確認する。
export function listVotedElections(regionId) {
  const state = loadState()
  return state.vote_records
    .filter((r) => r.region_id === regionId)
    .map((r) => ({ ...getElection(r.election_id, state), voted_at: r.voted_at }))
    .sort((a, b) => (a.voted_at < b.voted_at ? 1 : -1))
}

File: app/vote_records.py
"""EPIC-07 投票済みを記録できる。"""
import sqlite3
from datetime import datetime, timezone

from app.elections import get_election
from app.regions import get_region


def mark_voted(conn: sqlite3.Connection, region_id: int, election_id: int) -> dict:
    """「投票した」と記録する。同じ選挙への記録は冪等（二重記録しない）。"""
    get_region(conn, region_id)  # 存在確認
    get_election(conn, election_id)  # 存在確認
    now = datetime.now(timezone.utc).isoformat()
    conn.execute(
        "INSERT OR IGNORE INTO vote_records (region_id, election_id, voted_at) VALUES (?, ?, ?)",
        (region_id, election_id, now),
    )
    conn.commit()
    row = conn.execute(
        "SELECT * FROM vote_records WHERE region_id = ? AND election_id = ?",
        (region_id, election_id),
    ).fetchone()
    return dict(row)


def list_voted_elections(conn: sqlite3.Connection, region_id: int) -> list[dict]:
    """投票済みの選挙・過去の投票履歴を確認する。"""
    rows = conn.execute(
        "SELECT vote_records.voted_at, elections.* FROM vote_records "
        "JOIN elections ON elections.id = vote_records.election_id "
        "WHERE vote_records.region_id = ? ORDER BY vote_records.voted_at DESC",
        (region_id,),
    ).fetchall()
    return [dict(r) for r in rows]
